hw/ip/otbn/dv/rig/rig/gens/bad_load_store.py

```diff
@@ -0,0 +1,239 @@
+                               'generator.')
+
+        for insn in insns_file.insns:
+            if (insn.mnemonic in ['lw', 'sw', 'bn.lid', 'bn.sid']):
+                weight = cfg.insn_weights.get(insn.mnemonic)
+                if weight > 0:
+                    self.weights.append(weight)
+                    self.insns.append(insn)
+
+        # Check that at least one instruction has a positive weight
+        assert len(self.insns) == len(self.weights)
+        if not self.weights:
+            raise ValueError('Config at {} defines a zero weight '
+                             'for all instructions.'
+                             .format(cfg.path))
+
+    def gen(self,
+            cont: GenCont,
+            model: Model,
+            program: Program) -> Optional[GenRet]:
+
+        # Pick grd randomly. Since we can write to any register (including x0)
+        # this should always succeed. 
+        op_val_grd = model.pick_operand_value(self.grd_op_type)
+        assert op_val_grd is not None
+
+        # Determine the offset range for XW
+        offset_rng = self.imm_op_type.get_op_val_range(model.pc)
+        assert offset_rng is not None
+
+        # Max/Min Offsets for XW (-2048, 2047)
+        max_offset = offset_rng[1]
+        min_offset = offset_rng[0]
+
+        # Determine the offset range for BN.XID
+        bn_offset_rng = self.bn_imm_op_type.get_op_val_range(model.pc)
+        assert bn_offset_rng is not None
+
+        # Max/Min Offsets for BN.XID (-512 * 32, 511 * 32)
+        bn_max_offset = bn_offset_rng[1]
+        bn_min_offset = bn_offset_rng[0]
+
+        # Pick a really negative or really positive known value for hitting OOB
+        base = []
+
+        # Get known registers that has out of bounds addresses (from both sides)
+        known_regs = model.regs_with_known_vals('gpr')
+        for reg_idx, reg_val in known_regs:
+            if (min_offset + (1 << 32)) > reg_val > (4096 + min_offset):
+                base.append((reg_idx, reg_val))
+        if not base:
+            return None
+
+        # Pick a random register among known registers that is constrained above
+        idx, value = random.choice(base)
+
+        # If it is close enough, pick top of the memory for "barely" out of bound
+        # coverage point for XW instructions
+        if (4096 + min_offset) < value <= 4096 + max_offset:
+            imm_val = 4096 - value
+        else:
+            imm_val = random.randrange(min_offset, max_offset)
+
+        offset_val = self.imm_op_type.op_val_to_enc_val(imm_val, model.pc)
+        assert offset_val is not None
+
+        # It does not matter for BN.XID instructions
+        bn_imm_val = random.randrange(bn_min_offset, bn_max_offset, 32)
+        bn_offset_val = self.bn_imm_op_type.op_val_to_enc_val(bn_imm_val,
+                                                              model.pc)
+        assert bn_offset_val is not None
```

Comment:
Could we restructure things so that we only pick `imm_val` / `bn_imm_val` if we're generating the relevant instruction? I'm not sure I quite understand the logic here, but it would probably be easier to follow if there were `fill_xw_insn` / `fill_bn_insn` methods or something.